### ISSUE TITLE
temperature > 0 is silently deterministic on the second call onward — categorical_sampling's @mx.compile(outputs=mx.random.state) doesn't advance the PRNG

### ISSUE THREAD

**Issue by BlakeFlorida** [ml-explore/mlx-lm#1439]
Under inference with mlx-lm (and mlx-vlm, which reuses
`mlx_lm.sample_utils`), `temperature > 0` produces **byte-identical
output across repeat calls within a single process**, regardless of
`mx.random.seed()`, the `seed=` kwarg to `generate_step`, or any
sampling parameters.

Sampling appears to work for the first ~2 token positions in a process,
then silently collapses to a deterministic "stuck" sampler. Each token
position thereafter selects whichever token contains the same fixed
percentile of its distribution — confident decisions look greedy, less
confident decisions look weirdly off-distribution.

Server bounces are the only way to change output. Different processes
get different "stuck" sampler states, so behavior varies across
restarts but never within a session.

## Versions

- mlx-lm 0.31.3
- mlx-vlm 0.6.3 (also affected — reuses `mlx_lm.sample_utils`)
- mlx 0.31.2
- Python 3.12, macOS Darwin, Apple Silicon M-series

## Evidence

Two `/v1/chat/completions` calls, same prompt, `temperature=1.0`:

```
Request 1:  md5(content) = 7ac0dba2  "The user wants a haiku about coffee..."
Request 2:  md5(content) = 7ac0dba2  "The user wants a haiku about coffee..."
            ^^^^^^^^^^^^^ byte-identical
```

Tracing `mx.random.state` at each `categorical_sampling` call inside
the server shows the state advances once, then freezes:

```
call 1 state: [array([0, 2129127115])]                  ← initial
call 2 state: [array([1131398383, 4059925319])]         ← advanced
call 3 state: [array([1131398383, 4059925319])]         ← stuck
call 4 state: [array([1131398383, 4059925319])]         ← stuck
call 5 state: [array([1131398383, 4059925319])]         ← stuck
call 6 state: [array([1131398383, 4059925319])]         ← stuck
```

Same calls in a standalone Python script (no inference, just
`categorical_sampling(fake_logits, 1.0)` in a loop) DO advance state
correctly — so the bug only manifests in a real generation pipeline.
Something about `generate_step`'s context (compiled forward pass,
multiple `mx.compile`-decorated helpers sharing state, `mx.stream`
nesting, possibly thread affinity in a single-worker threadpool)
triggers the freeze.

## Workaround (confirmed working)

Replace `categorical_sampling` with a non-compiled version at process
startup:

\`\`\`python
import mlx.core as mx
import mlx_lm.sample_utils as su

def _cat_uncompiled(logits, temp):
    return mx.random.categorical(logits * (1 / temp))

su.categorical_sampling = _cat_uncompiled

# mlx-vlm imports the same symbol — patch it too if used:
try:
    import mlx_vlm.sample_utils as vsu
    vsu.categorical_sampling = _cat_uncompiled
except ImportError:
    pass
\`\`\`

After this patch:
- Same prompt + same `temperature > 0` → varied output across calls ✓
- `mx.random.seed()` actually changes output ✓
- Reasoning models with thinking blocks now produce varied actions, not
  just varied reasoning text

## Why this matters

CoT / thinking-mode models are particularly sensitive. They produce
varied `reasoning_content` (the thinking-phase sampler hits a
different code path) but byte-identical action/content output. So
the model "thinks" differently each turn while always emitting the
same final answer — sampling-driven exploration is silently lost.

For tool-calling agents this manifests as unbreakable loops: the model
re-thinks its decision each turn but always emits the same broken tool
call because the action distribution is peaked and the stuck percentile
always lands on the same top token.

For standard chat use, output across sessions feels "inconsistent"
(each process restart picks a different stuck state) while output
within a session is suspiciously deterministic. Easy to misattribute
to model variance or quantization noise.

## Suggested fix

The most direct option: remove the `@partial(mx.compile, ...)`
decorator from `categorical_sampling`. The body is a single
`mx.random.categorical` call — compile gains are negligible and the
correctness cost is total.

Longer-term: investigate whether `mx.compile` with
`outputs=mx.random.state` is supposed to bind state on first invocation
under heavy inference contexts. If yes, document the contract; if no,
that's an mlx-level issue worth filing separately.

## Related code

- `mlx_lm/sample_utils.py` — `categorical_sampling` decorator
- `mlx_lm/generate.py` (and `mlx_vlm/generate/`) — sampler call sites
  during decode loop
- `mlx_vlm/sample_utils.py` — re-exports the affected symbol

**Comment by jax-0n-git** [comment]
Confirmed on 0.31.3 / mlx 0.31.2 (Python 3.14 venv, M5 Max, macOS 26.5.1), via the `mlx_lm.server` batch path — and a data point on how silently this bites: a 206-generation benchmark (40 prompts × 5 samples, `temperature: 1.0`) came back with **all 40 prompts having 5/5 byte-identical samples** — effective sample size collapsed 200→40 with no error anywhere. Request-body `seed` is also inert in this state (`seed: 1` and `seed: 2` return identical completions).

Minimal workaround verified: commenting the `@partial(mx.compile, inputs=mx.random.state, outputs=mx.random.state)` decorator off `categorical_sampling` in `sample_utils.py` restores per-call stochasticity (3 identical temp-1.0 requests → 3 distinct outputs immediately after a server restart). Eager `mx.random.categorical` advances the global state fine — it's specifically the compiled wrapper that freezes it, matching your trace.

Also reproduces through a downstream distribution that vendors mlx_lm 0.31.3 (vMLX 1.5.54, one of its two serving paths), so the blast radius includes bundled apps.
